docs/001-mana_proposal.md

```diff
@@ -0,0 +1,539 @@
+other way around.
+
+### Initialization
+
+Mana state machine is an extension of the ledger state, hence its calculation depends on the ledger state perception
+of the node. Snapshotting is the mechanism that saves the ledger states and prunes unnecessary transactions. Together
+with the ledger state, base mana vectors are also saved, since a certain ledger state reflects a certain mana distribution
+in the network. In future, when snapshotting is implemented in GoShimmer, nodes joining the network will be able to query
+for snapshot files that will contain initial base mana vectors as well.
+
+Until this functionality is implemented, mana calculation solely relies on transactions getting confirmed. That is, when
+a node joins the network and starts gathering messages and transactions from peers, it builds its own ledger state through
+solidification process. Essentially, the node requests all messages down to the genesis from the current tips of its neighbors.
+Once the genesis is found, messages are solidified bottom up. For the value tangle, this means that for each solidified
+and liked transaction, `TransactionConfirmed` event is triggered, updating the base mana vectors.
+
+In case of a large database, initial synching and solidification is a computationally heavy task due to the sheer amount
+of messages in the tangle. Mana calculation only adds to this burden. It will be determined through testing if additional
+"weight lifting" mechanism is needed (for example delaying mana calculation).
+
+In Pollen test network, all funds are initially held by the faucet node, therefore all mana present at bootstrap belong
+to this node. Whenever a transaction is requested from the faucet, it pledges mana to the requesting node, helping other
+nodes to increase their mana.
+
+### Mana Package
+
+The functionality of the mana module should be implemented in a `mana` package. Then, a `mana plugin` can use the package
+structs and methods to connect the dots, for example execute `BookMana` when `TransactionConfirmed` event is triggered
+in the value tangle.
+
+`BaseMana` is a struct that holds the different mana values for a given node.
+Note that except for `Base Mana 1` calculation, we need the time the `BaseMana` values were updated, so we store it in the struct:
+ ```go
+type BaseMana struct {
+  BaseMana1 float
+  EffectiveBaseMana1 float
+  BaseMana2 float
+  EffectiveBaseMana2 float
+  LastUpdated time
+}
+```
+
+`BaseManaVector` is a data structure that maps `nodeID`s to `BaseMana`.
+```go
+type BaseManaVector map[nodeID]BaseMana
+```
+
+#### Methods
+`BaseManaVector` should have the following methods:
+ - `BookMana(transaction)`: Book mana of a transaction. Trigger `ManaBooked` event. Note, that this method updates
+   `BaseMana` with respect to time and to new `Base Mana 1` and `Base Mana 2` values.
+ - `GetTotalMana(nodeID) mana`: Return `Effective Base Mana 1` + `Effective Base Mana 2` of a particular node. Note, that
```

Comment:
Why we would need the sum of Mana1 and Mana2?

Reply:
Right, let's change it like this:
```go
// GetWeightedMana returns the combination of Effective Base Mana 1 & 2, weighted by weight.
// mana = EBM1 * weight + EBM2 * ( 1- weight), where weight is in [0,1].
func (bmv *BaseManaVector) GetWeightedMana(nodeID identity.ID, weight float64) (float64, error) {
	if _, exist := bmv.vector[nodeID]; !exist {
		return 0.0, errors.New("node not present in base mana vector")
	}
	if weight < 0.0 || weight > 1.0 {
		return 0.0, errors.New("invalid weight parameter, outside of [0,1]")
	}
	bmv.Update(nodeID, time.Now())
	baseMana := bmv.vector[nodeID]
	return baseMana.EffectiveBaseMana1*weight + baseMana.EffectiveBaseMana2*(1-weight), nil
}

// GetMana returns the 50 - 50 split combination of Effective Base Mana 1 & 2.
func (bmv *BaseManaVector) GetMana(nodeID identity.ID) (float64, error) {
	return bmv.GetWeightedMana(nodeID, 0.5)
}
```